File: 042.zig
const std = @import("std");
const stdout = std.io.getStdOut().writer();
const words = @embedFile("./resources/0042_words.txt");
const sqrt = std.math.sqrt;

fn getWordValue(word: []const u8) u64 {
    var value: u64 = 0;
    for (word) |c| {
        value += c - 'A' + 1;
    }

    return value;
}

fn isTriangular(n: u64) bool {
    const x = 8*n + 1;
    return x % sqrt(x) == 0;
}

pub fn main() !void {
    var word_it = std.mem.splitAny(u8, words, ",");
    var triangle_words: u64 = 0;

    while (word_it.next()) |word| {
        if (isTriangular(getWordValue(word[1..word.len - 1]))) {
            triangle_words += 1;
        }
    }

    try stdout.print("{}\n", .{triangle_words});
}

File: 042.c
// 162
#include <stdio.h>
#include <string.h>
#include <math.h>

#define N_WORDS 1786
#define LINE_SZ 16346

int get_word_value(const char *word)
{
    int value = 0;
    size_t word_len = strlen(word);

    for (int i = 1; i < word_len - 1; i++) { // ignore quotes
        value += word[i] - 64;
    }

    return value;
}

int is_triangular(int n)
{
    double n0 = 2.0*n;
    int int_part = sqrt(n0);
    int_part *= (int_part + 1);

    return int_part/2 == n;
}

int main()
{
    char line[LINE_SZ] = { 0 };
    const char *words[N_WORDS] = {};
    char *tok = NULL;
    FILE *f = fopen("resources/0042_words.txt", "r");
    int n = 0;
    const char *delim = ",";

    fgets(line, LINE_SZ, f);
    tok = strtok(line, delim);

    // store words
    for (int i = 0; tok != NULL; i++) {
        words[i] = tok;
        tok = strtok(NULL, delim);
    }

    // count how many words are triangle words
    for (int i = 0; i < N_WORDS; i++) {
        int value = get_word_value(words[i]);
        n += is_triangular(value);
    }

    printf("%d\n", n);

    fclose(f);
    return 0;
}
